[CGP-370] Added type checking rules, fixed a bunch of missing-kind-checking bugs

effectfully opened the pull request:
More work on the type checker. This PR adds type checking rules, a bit of comments and fixes several bugs and an incorrect golden test related to missing kind checking. For convenient kind checking I added the `kindCheckM` function and also simplified `kindOf` using it.

This doesn't touch constants, because another my PR deals with them. Once both are merged, I'll do appropriate changes.

vmchale commented on language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs:
```diff
@@ -155,43 +151,36 @@ typeCheckStep = do
+kindCheckM :: a -> Type TyNameWithKind a -> Kind () -> TypeCheckM a ()
+kindCheckM x ty k = do
+    tyK <- kindOf ty
+    when (tyK /= k) . throwError $ KindMismatch x (void ty) k tyK
```
The `when p . throwError` might be simpler as `when p $ throwError` or something else.

effectfully replied:
Agreed.

vmchale commented on language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs:
```diff
@@ -219,23 +208,38 @@ dummyType = TyVar () dummyTyName
+    -- ---------------------------------------------------------
+    -- [infer| lam n dom body : vDom -> vCod]
+
+    kindCheckM x dom $ Type ()
```
The type inference rules for `(lam y T M)` do not currently require that `T` have kind `(type)`. Is this something we discussed? The spec should be updated if this is the case.

effectfully replied:
The rule in the spec is

```
Γ, y : T |- M : S
--------------------
Γ |- (lam y T M ) : (fun T S)
```

`fun T S` is simply ill-kinded if `T` is not of kind `*`. But maybe the spec should indeed say

> In context Γ, term M has normal well-kinded type S

instead of

> In context Γ, term M has normal type S

Just for clarity. @psygnisfive, what do you think?

mchakravarty replied:
@effectfully Yes I agree that this check should be part of the hypotheses (and hence, is appropriate in the code).

effectfully replied:
Then I'll add a bit more docs and merge. Thanks.

mchakravarty replied:
Yes, I agree that it is best to update the spec and implementation in tandem.

effectfully replied:
While I think it's nice to keep the spec in sync with the implementation, that would also slow down the development process considerably. Changing the spec (and any theoretical work in general) is slow by its nature. And we want development to be fast.

Regarding this particular issue: it's at worst a documentation problem in the spec. No updates to theory are required. And the plutus-metatheory repo handles everything correctly. It is also obviously correct that the kind of the type of the variable a lambda binds should be `*`, because only types of this kind are inhabited by terms. So I do not want this PR to be blocked on this.

mchakravarty replied:
To further clarify, I think, it has been a flaw in our process that the spec changes have been so slow going. One goal of having the spec alongside the code in the same repo is to change this.

If you are unsure of how to change the LaTeX, please let me know and I'll add the diff.

effectfully replied:
I'm very reluctant to change the spec myself. I'll add those few symbols to the spec, but
1) I'm not confident they're needed there
2) there could be a better description in a different place

effectfully replied:
Done. Also this introduces a merge conflict with the recent PR by @psygnisfive.

effectfully replied:
> And merge conflicts are a fact of life if we want to make good progress as a team of our current size.

Sure, I'm just CC'ed Rebecca, so she knows where the conflict comes from.

> I was actually thinking of also extending the inference rule to

I added kind-checking to `Error`, `TyInst` and `Wrap` as well, so these cases should be changed too. I'll start a conversation about this in slack.

psygnisfive replied:
Ahhhh interesting. I'm going to add the following judgment-defining rules to the spec:

`--------------
Γ ⊢ α :: K jdg`

`Γ ⊢ A :: type`
-------------
Γ ⊢ x : A jdg`

Does this make sense to everyone? Or should I make it slightly less formal b/c of the risk of weird circularity?

vmchale replied:
Is there a formal approach that avoids circularity?

vmchale commented on language-plutus-core/test/types/reduce.plc:
```diff
@@ -1,5 +1,5 @@
 (program 0.1.0
-  (lam x [(lam x (type) x) [(lam y (type) y) (con integer)]]
```
Is this actually invalid according to the spec?

effectfully replied:
This is the same issue as above, so I'm resolving it.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs:
```diff
@@ -219,23 +208,38 @@ dummyType = TyVar () dummyTyName
+typeOf (TyAbs _ n nK body)                       =
+    -- [infer| body : vBodyTy]
+    -- ----------------------------------------------
+    -- [check| abs n nK body : all (n :: nK) vBodyTy]
```
A thinko: this should clearly be `infer`.

effectfully commented:
> I think, this is the more conventional formatting (and the style, eg, GHC tends to use). What do you think?

I tried doing this. But that blank like between clauses looked kinda ugly to me, because blank lines is what we use to separate definitions from each other. I think what we have there right now reads slightly easier. But that's not important at all and I can reformat the code.

I guess, I'll reformat the code, we'll look at the result and then decide.

mchakravarty commented:
> I tried doing this. But that blank like between clauses looked kinda ugly to me, because blank lines is what we use to separate definitions from each other.

I understand, we usually hesitate to separate equations, but in long multiline equations, especially if there are many of them (which is a common situation when traversing an AST in a compiler), I think, it makes sense to adopt this style (and, as I mentioned, there is some precedent and hence familiarity among compiler writers in Haskell).

effectfully commented:
Comments addressed.

I added some notes about the notation used and also added the `typeOfConstant` function to simplify `typeOf`.

> I understand, we usually hesitate to separate equations, but in long multiline equations, especially if there are many of them (which is a common situation when traversing an AST in a compiler), I think, it makes sense to adopt this style (and, as I mentioned, there is some precedent and hence familiarity among compiler writers in Haskell).

Do not have any strong feelings, but I liked the previous version slightly more. That's a minor thing.

mchakravarty commented:
Check out https://github.com/ghc/ghc/blob/master/compiler/typecheck/TcExpr.hs There are no rules at the equations, but this modules also uses the style with empty lines between equations heavily and IMHO it works very well.

effectfully commented:
> Check out https://github.com/ghc/ghc/blob/master/compiler/typecheck/TcExpr.hs There are no rules at the equations, but this modules also uses the style with empty lines between equations heavily and IMHO it works very well.

It looks much better than when there are equations.

mchakravarty commented:
What is with the CI checks. Are they hanging again?

effectfully commented:
> What is with the CI checks. Are they hanging again?

No idea. @iohk-devops, @disassembler, do you know what is with CI?

mchakravarty commented:
@effectfully Did the checks pass before you pushed the spec change?

effectfully commented:
> @effectfully Did the checks pass before you pushed the spec change?

I normally do not run all the checks locally, because this requires building, testing and benchmarking everything, which just take too long when you need only a quick sanity check, so I delegate this work to CI. I checked right now, seems fine to me.

effectfully commented:
(I do not have the rights required to perform the merge myself)

effectfully commented:
Err, this one should have been squash-merged.

wadler commented:
I don't understand these rules at all. Can you explain Rebecca? P

On Thu, 11 Oct 2018 at 07:18, Rebecca Valentine <notifications@github.com>
wrote:

> *@psygnisfive* commented on this pull request.
> ------------------------------
>
> In language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs
> <https://github.com/input-output-hk/plutus/pull/179#discussion_r224325996>
> :
>
> >  -- order to maintain global uniqueness.
>  --
>  -- This is all tracked in a state monad, and we simply delete any substitutions
>  -- once they go out of scope; this is permissible since 'Unique's are globally
>  -- unique and so we will not delete the wrong thing.
>  -- | Synthesize the type of a term, returning a normalized type.
>  typeOf :: Term TyNameWithKind NameWithType a -> TypeCheckM a (NormalizedType TyNameWithKind ())
> -typeOf (Var _ (NameWithType (Name (_, ty) _ _))) = normalizeTypeOpt $ void ty
> -typeOf (LamAbs _ _ ty t)                         =
> -    TyFun () <<$>> normalizeTypeOpt (void ty) <<*>> typeOf t
> +typeOf (Var _ (NameWithType (Name (_, ty) _ _))) =
> +    -- Since we kind check types at lambdas, we can normalize types here without kind checking.
> +    -- Type normalization at each variable is inefficient and we may consider something else later.
> +    normalizeTypeOpt $ void ty
> +typeOf (LamAbs x _ dom body)                     = do
> +    -- [check| dom :: *]    dom ~>? vDom    [infer| body : vCod]
> +    -- ---------------------------------------------------------
> +    -- [infer| lam n dom body : vDom -> vCod]
> +
> +    kindCheckM x dom $ Type ()
>
> Ahhhh interesting. I'm going to add the following judgment-defining rules
> to the spec:
>
> -------------- Γ ⊢ α :: K jdg
> Γ ⊢ A :: type
>
> Γ ⊢ x : A jdg`
>
> Does this make sense to everyone? Or should I make it slightly less formal
> b/c of the risk of weird circularity?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/input-output-hk/plutus/pull/179#discussion_r224325996>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEqfshgjVFe8XT04HidRnsqhIFdWrPelks5ujuK6gaJpZM4XNYwS>
> .
>
-- 
.   \ Philip Wadler, Professor of Theoretical Computer Science,
.   /\ School of Informatics, University of Edinburgh
.  /  \ and Senior Research Fellow, IOHK
. http://homepages.inf.ed.ac.uk/wadler/

Too brief? Here's why: http://www.emailcharter.org/

The University of Edinburgh is a charitable body, registered in
Scotland, with registration number SC005336.

effectfully commented:
@psygnisfive, that's a nice idea:

```
--------------
Γ ⊢ α :: K jdg

Γ ⊢ A :: (type)
---------------
Γ ⊢ x : A jdg
```

`jdg` is a meta-judgement that defines judgements, right?

However I think this is usually kept implicit like `Γ valid`, for example. I've never seen meta-judgements being used, but the idea is kinda natural, so maybe we could find a reference?

mchakravarty commented:
@effectfully 

> Err, this one should have been squash-merged.

Did we agree to do this now?

effectfully commented:
> Did we agree to do this now?

Yes, sorry, I just forgot to say about squashing.
